extras/sht3x/sht3x.c

```diff
@@ -0,0 +1,419 @@
+{
+    if (!dev || !dev->active) return false;
+
+    uint8_t data[2] = { cmd >> 8, cmd & 0xff };
+
+    debug_dev ("send command MSB=%02x LSB=%02x", __FUNCTION__, dev, data[0], data[1]);
+
+    int err;
+    int count = 10;
+
+    // in case i2c is busy, try to write up to ten times and 100 ms
+    // tested with a task that is disturbing by using i2c bus almost all the time
+    while ((err=i2c_slave_write(dev->bus, dev->addr, 0, data, 2)) == -EBUSY && count--)
+       vTaskDelay (10 / portTICK_PERIOD_MS);
+  
+    if (err)
+    {
+        dev->error_code |= (err == -EBUSY) ? SHT3x_I2C_BUSY : SHT3x_I2C_SEND_CMD_FAILED;
+        error_dev ("i2c error %d on write command %02x", __FUNCTION__, dev, err, cmd);
+        return false;
+    }
+      
+    return true;
+}
+
+
+static bool sht3x_read_data(sht3x_sensor_t* dev, uint8_t *data,  uint32_t len)
+{
+    if (!dev || !dev->active) return false;
+
+    int err;
+    int count = 10;
+
+    // in case i2c is busy, try to read up to ten times and 100 ms
+    while ((err=i2c_slave_read(dev->bus, dev->addr, 0, data, len)) == -EBUSY && count--)
+        vTaskDelay (10 / portTICK_PERIOD_MS);
+        
+    if (err)
+    {
+        dev->error_code |= (err == -EBUSY) ? SHT3x_I2C_BUSY : SHT3x_I2C_READ_FAILED;
+        error_dev ("error %d on read %d byte", __FUNCTION__, dev, err, len);
+        return false;
+    }
+
+#   ifdef SHT3x_DEBUG
+    printf("SHT3x %s: bus %d, addr %02x - read following bytes: ", 
+           __FUNCTION__, dev->bus, dev->addr);
+    for (int i=0; i < len; i++)
+        printf("%02x ", data[i]);
+    printf("\n");
+#   endif
+
+    return true;
+}
+
+
+/*
+static bool sht3x_reset (sht3x_sensor_t* dev)
+{
+    if (!dev || !dev->active) return false;
+
+    debug_dev ("soft-reset triggered", __FUNCTION__, dev);
+    
+    dev->error_code = SHT3x_OK;
+
+    // send reset command
+    if (!sht3x_send_command(dev, SHT3x_RESET_CMD))
+    {
+        dev->error_code |= SHT3x_SEND_RESET_CMD_FAILED;
+        return false;
+    }   
+    // wait for small amount of time needed (according to datasheet 0.5ms)
+    vTaskDelay (20 / portTICK_PERIOD_MS);
+    
+    uint16_t status;
+
+    // check the status after reset
+    if (!sht3x_get_status(dev, &status))
+        return false;
+        
+    return true;    
+}
+*/
+
+
+static bool sht3x_get_status (sht3x_sensor_t* dev, uint16_t* status)
+{
+    if (!dev || !dev->active || !status) return false;
+
+    dev->error_code = SHT3x_OK;
+
+    uint8_t  data[3];
+
+    if (!sht3x_send_command(dev, SHT3x_STATUS_CMD) || !sht3x_read_data(dev, data, 3))
+    {
+        dev->error_code |= SHT3x_STATUS_CMD_FAILED;
+        return false;
+    }
+
+    *status = data[0] << 8 | data[1];
+    debug_dev ("status=%02x", __FUNCTION__, dev, *status);
+    return true;
+}
+
+
+
+static uint8_t crc8_table[256];         // lookup table with precomputed crc values
+static bool crc8_first_time = true;     // indicator whether table has still to be created
```

Comment:
Would need to make this `const` to have it placed in flash. I suggest using a non-table implementation. A quick look suggests the following might work, or be close:

```
static uint8_t sht3x_crc8(uint8_t data[], uint8_t num_bytes)
{
    uint8_t crc = 0xff;
    uint8_t i;

    for (i = 0; i < num_bytes; i++) {
        crc ^= data[i];
        for (uint8_t bit = 8; bit > 0; bit++) {
            if (crc & 0x80)
                crc = (crc << 1) ^ 0x31;
            else
                crc = (crc << 1);
        }
    }
    return crc;
}
```

Reply:
Yes, implementation is almost the same. I used the lookup table version because it is less time consuming. Ok, if it is used only one time per second or less, computation time doesn't matter.

I will change it.